### ISSUE TITLE
TOB-K8S-032: CoreDNS leaks internal cluster information across namespaces

### ISSUE THREAD

**Issue by cji** [kubernetes/kubernetes#81137]
This issue was reported in the [Kubernetes Security Audit Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Description**
Kubernetes can use DNS for service discovery. This allows both cluster components and containers alike to retrieve service and other information from a single source. However, CoreDNS does not distinguish between clients requesting information, allowing attackers to gain access to information outside of the current namespace.

```
root@wordpress-dccb8668f-mzg45:/var/www/html# nslookup -type=ns default.svc.cluster.local
;; Truncated, retrying in TCP mode.
Server:		10.233.0.3
Address:	10.233.0.3#53

cluster.local
	origin = ns.dns.cluster.local
	mail addr = hostmaster.cluster.local
	serial = 1555691051
	refresh = 7200
	retry = 1800
	expire = 86400
	minimum = 30
wordpress.default.svc.cluster.local	service = 0 100 80 wordpress.default.svc.cluster.local.
_http._tcp.wordpress.default.svc.cluster.local	service = 0 100 80 wordpress.default.svc.cluster.local.
kubernetes-dashboard.kube-system.svc.cluster.local	service = 0 100 443 kubernetes-dashboard.kube-system.svc.cluster.local.
...
_metrics._tcp.coredns.kube-system.svc.cluster.local	service = 0 100 9153 
...
```
Figure 28.1: An attacker querying name server (NS) records from a compromised container

**Exploit Scenario**
Alice has configured a cluster with multiple namespaces, so as to segment her containers across multiple units of her organization. Eve has access to a compromised container, and issues a name server query to CoreDNS in order to further map hosts and services for potential compromise. 

**Recommendation**
Short term, document this behavior so that users and administrators are aware that CoreDNS may leak information in specific configurations.

Long term, work with CoreDNS to implement per-namespace DNS information, so as to minimize the leak of information across namespaces. This will allow administrators to denote which information is shared across environments, and prevent mistakes from cascading across namespaces or attackers from gaining access to information surrounding multiple namespaces across the cluster. 

**Anything else we need to know?**:

See #81146 for current status of all issues created from these findings.

The vendor gave this issue an ID of TOB-K8S-032 and it was finding 29 of the report.

The vendor considers this issue Low Severity.

To view the original finding, begin on page 73 of the [Kubernetes Security Review Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Environment**:

- Kubernetes version: 1.13.4

**Comment by neolit123** [comment]
i do not see a coredns version in this report.
possibly the one kubespray->kubeadm uses (as per the summary of the audit) which is 1.3.1

cc @chrisohaver @rajansandeep

**Comment by joelsmith** [comment]
I don't think this is an issue with CoreDNS. Rather, because our use of CoreDNS conforms to the Kubernetes DNS spec, this discovery is possible.

https://github.com/kubernetes/dns/blob/master/docs/specification.md

Any DNS implementation that serves SRV records as described would have the same behavior.

As long as we have ways to disable DNS-based service discovery and as long as we document the security implications of using DNS-SD, we're probably okay to close this issue.

**Comment by chrisohaver** [comment]
I don't think this is behavior that is within spec, but I could be misunderstanding.  Per the report, this is performed in a K8s 1.13.1 cluster.

I _can_ replicate this issue in CoreDNS 1.2.6 (released with k8s 1.13).
I _cannot_ replicate it in CoreDNS 1.3.1 (released with k8s 1.14 and 1.15).

In CoreDNS 1.2.6, it appears that the NS query triggers a zone transfer to the client.
In CoreDNS 1.13, the behavior seems to be prevented by the fix in coredns/coredns#2292, although I have not carefully verified.

**Comment by joelsmith** [comment]
@chrisohaver you're right. I didn't read the finding carefully enough and I thought it was referring to the SRV records used for DNS-SD.

**Comment by chrisohaver** [comment]
Update: 

I verified that in 1.2.6, there was a bug in the kubernetes plugin zone transfer feature, that always enabled zone transfers to all clients by default.  This was fixed in 1.3.1 with coredns/coredns#2292.  As of 1.3.1, zone transfers are only permitted if enabled, and client permitted in the Corefile.

I just opened coredns/coredns#3098 to fix the issue where _if zone transfers are enabled_ in the CoreDNS kubernetes plugin, an `NS` request for an arbitrary subdomain of the cluster zone would trigger a zone transfer of the cluster zone.

Note: Zone transfers are not part of the [dns spec](https://github.com/kubernetes/dns/blob/master/docs/specification.md). As of CoreDNS 1.3.1, zone transfers are not enabled by default.

**Comment by neolit123** [comment]
/lifecycle frozen
@chrisohaver do we need to continue keeping this issue open?

**Comment by rajansandeep** [comment]
According to https://github.com/kubernetes/kubernetes/issues/81137#issuecomment-519572531, the issue is fixed in CoreDNS v1.6.2 which was released with k8s 1.16, so I think it's safe to say this can be closed.

**Comment by neolit123** [comment]
thanks @rajansandeep 
/close
